syntax = "proto3";

package kvstore;

// Request message for the Put operation
message PutRequest {
    string key = 1;
    string value = 2;
    int64 version = 3;
}

// Response message for the Put operation
message PutResponse {
    int64 version = 1;
    bool success = 2;
}

// Request message for the Get operation
message GetRequest {
    string key = 1;
}

// Response message for the Get operation
message GetResponse {
    string value = 1;
    int64 version = 2;
    bool found = 3;
}

// Request message for the Delete operation
message DeleteRequest {
    string key = 1;
}

// Response message for the Delete operation
message DeleteResponse {
    bool success = 1;
}
// Heartbeat Request message
message HeartbeatRequest {}

// Heartbeat Response message
message HeartbeatResponse {}

// Service definition
service KVStoreRPC {
    rpc Put(PutRequest) returns (PutResponse);
    rpc Get(GetRequest) returns (GetResponse);
    rpc Del(DeleteRequest) returns (DeleteResponse);
    rpc SendHeartbeat(HeartbeatRequest) returns (HeartbeatResponse); // New RPC for heartbeat
}
